File: 3d/src/main/java/io/github/xerooup/frame3d/core/internal/Engine.java
package io.github.xerooup.frame3d.core.internal;

import io.github.xerooup.frame3d.core.Game;
import io.github.xerooup.frame3d.core.Settings;
import io.github.xerooup.frame3d.graphics.Color;
import io.github.xerooup.frame3d.graphics.internal.RenderContextImpl;
import org.lwjgl.glfw.GLFWErrorCallback;
import org.lwjgl.opengl.GL;

import static org.lwjgl.glfw.Callbacks.glfwFreeCallbacks;
import static org.lwjgl.glfw.GLFW.*;
import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.system.MemoryUtil.NULL;

public class Engine {
    private final Settings settings;
    private final Game game;
    private long window;
    private boolean running = false;
    private int frameCount = 0;

    private static Engine instance;

    public Engine(Settings settings, Game game) {
        this.settings = settings;
        this.game = game;
    }

    public static void forceStop() {
        if (instance != null) {
            instance.running = false;
        }
    }

    public void run() {
        instance = this;

        try {
            init();
            game.create();
            loop();
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            cleanup();
        }
    }

    private void init() {
        GLFWErrorCallback.createPrint(System.err).set();

        if (!glfwInit()) {
            throw new IllegalStateException("unable to initialize GLFW");
        }

        glfwDefaultWindowHints();
        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
        glfwWindowHint(GLFW_RESIZABLE, settings.resizable ? GLFW_TRUE : GLFW_FALSE);
        glfwWindowHint(GLFW_DECORATED, settings.decorated ? GLFW_TRUE : GLFW_FALSE);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GLFW_TRUE);

        window = glfwCreateWindow(settings.width, settings.height, settings.title, NULL, NULL);
        if (window == NULL) {
            throw new RuntimeException("failed to create GLFW window");
        }

        if (settings.windowX != null && settings.windowY != null) {
            glfwSetWindowPos(window, settings.windowX, settings.windowY);
        } else {
            var vidmode = glfwGetVideoMode(glfwGetPrimaryMonitor());
            if (vidmode != null) {
                glfwSetWindowPos(window,
                        (vidmode.width() - settings.width) / 2,
                        (vidmode.height() - settings.height) / 2
                );
            }
        }

        glfwMakeContextCurrent(window);
        glfwSwapInterval(1);
        glfwShowWindow(window);

        GL.createCapabilities();

        glEnable(GL_DEPTH_TEST);
        glCullFace(GL_BACK);

        Color bg = settings.background;
        glClearColor(bg.r, bg.g, bg.b, 1f);
    }

    private void loop() {
        running = true;
        double startTime = glfwGetTime();
        double lastFpsTime = startTime;
        double lastTime = startTime;

        RenderContextImpl renderContext = new RenderContextImpl(settings.width, settings.height);

        while (!glfwWindowShouldClose(window) && running) {
            double currentTime = glfwGetTime();
            float delta = (float)(currentTime - lastTime);
            lastTime = currentTime;

            frameCount++;
            if (currentTime - lastFpsTime >= 1.0) {
                Game.fps = frameCount;
                frameCount = 0;
                lastFpsTime = currentTime;
            }

            Game.uptime = (float)(currentTime - startTime);

            game.update(delta);

            glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

            game.render(renderContext);

            glfwSwapBuffers(window);
            glfwPollEvents();
        }

        renderContext.cleanup();
    }

    private void cleanup() {
        try {
            if (window != 0) {
                glfwFreeCallbacks(window);
                glfwDestroyWindow(window);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        try {
            glfwTerminate();
        } catch (Exception e) {
            e.printStackTrace();
        }

        try (GLFWErrorCallback callback = glfwSetErrorCallback(null)) {
            if (callback != null) {
                callback.free();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: 2d/src/main/kotlin/io/github/xerooup/frame2d/core/internal/Engine.kt
package io.github.xerooup.frame2d.core.internal

import org.lwjgl.glfw.GLFW
import org.lwjgl.opengl.GL
import org.lwjgl.opengl.GL11
import org.lwjgl.system.MemoryUtil
import io.github.xerooup.frame2d.core.Game
import io.github.xerooup.frame2d.core.Settings
import io.github.xerooup.frame2d.graphics.DrawContext
import io.github.xerooup.frame2d.graphics.FrameBuffer
import io.github.xerooup.frame2d.graphics.internal.DrawContextImpl
import io.github.xerooup.frame2d.graphics.internal.TextureRenderer
import io.github.xerooup.frame2d.input.Keyboard
import io.github.xerooup.frame2d.input.Mouse

class Engine(private val settings: Settings, private val game: Game) {
    private var window: Long = 0
    private var lastFrameTime = 0L
    private var fpsCounter = 0
    private var fpsTimer = 0f
    private lateinit var frameBuffer: FrameBuffer
    private lateinit var drawContext: DrawContext
    private lateinit var textureRenderer: TextureRenderer

    companion object {
        private var currentInstance: Engine? = null

        fun forceStop() {
            currentInstance?.window?.let { window ->
                GLFW.glfwSetWindowShouldClose(window, true)
            }
        }
    }

    fun run() {
        currentInstance = this
        initWindow()
        initGL()
        initFrameBuffer()
        game.create()

        lastFrameTime = System.nanoTime()
        loop()
        cleanup()
        currentInstance = null
    }

    private fun initWindow() {
        if (!GLFW.glfwInit()) throw IllegalStateException("failed to initialize glfw")
        GLFW.glfwDefaultWindowHints()
        if (!settings.decorated) {
            GLFW.glfwWindowHint(GLFW.GLFW_DECORATED, GLFW.GLFW_FALSE)
            println(GLFW.GLFW_FALSE)
        }
        window = GLFW.glfwCreateWindow(
            settings.width,
            settings.height,
            settings.title,
            MemoryUtil.NULL,
            MemoryUtil.NULL
        )
        if (settings.windowX != null && settings.windowY != null)
            GLFW.glfwSetWindowPos(window, settings.windowX!!, settings.windowY!!)
        IconLoader.setIcon(window, settings.icon)
        Keyboard.init(window)
        Mouse.init(window)
        if (window == MemoryUtil.NULL) throw RuntimeException("failed to create window")
        GLFW.glfwMakeContextCurrent(window)
        GLFW.glfwSwapInterval(if (settings.targetFPS > 0) 1 else 0)
    }

    private fun initGL() {
        GL.createCapabilities()
        GL11.glClearColor(0f, 0f, 0f, 1f)
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT)
        GL11.glEnable(GL11.GL_TEXTURE_2D)
        GL11.glDisable(GL11.GL_DEPTH_TEST)

        // projection
        GL11.glMatrixMode(GL11.GL_PROJECTION)
        GL11.glLoadIdentity()
        GL11.glOrtho(0.0, settings.width.toDouble(), settings.height.toDouble(), 0.0, -1.0, 1.0)
        GL11.glMatrixMode(GL11.GL_MODELVIEW)
        GL11.glLoadIdentity()

        // blending
        GL11.glEnable(GL11.GL_BLEND)
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA)
    }

    private fun initFrameBuffer() {
        frameBuffer = FrameBuffer(settings.width, settings.height)
        drawContext = DrawContextImpl(frameBuffer)
        textureRenderer = TextureRenderer(frameBuffer)
    }

    private fun loop() {
        while (!GLFW.glfwWindowShouldClose(window)) {
            val currentTime = System.nanoTime()
            val delta = (currentTime - lastFrameTime) / 1_000_000_000f
            lastFrameTime = currentTime

            fpsCounter++
            fpsTimer += delta
            Game.uptime += delta

            if (fpsTimer >= 1f) {
                Game.fps = fpsCounter
                fpsCounter = 0
                fpsTimer = 0f
            }

            game.update(delta)

            Keyboard.update()
            Mouse.update()

            frameBuffer.clear(settings.background)
            game.render(drawContext)

            textureRenderer.updateTexture()
            textureRenderer.render()

            GLFW.glfwSwapBuffers(window)
            GLFW.glfwPollEvents()
        }
    }

    private fun cleanup() {
        textureRenderer.dispose()
        GLFW.glfwDestroyWindow(window)
        GLFW.glfwTerminate()
    }
}